url3472127e1f_SergeyKrivohatskiy_tiny_database_tgz-pJ8-ru_spbau_tinydb_bufferManager_IndexTestJ8/src/ru/spbau/tinydb/btree/IndexNode.java

```diff
@@ -0,0 +1,108 @@
+package ru.spbau.tinydb.btree;
+
+import java.util.Iterator;
+
+import ru.spbau.tinydb.bufferManager.BufferManager;
+
+public class IndexNode extends Node {
+
+	protected IndexNode(int pageIndex, BufferManager bm) {
+		super(pageIndex, bm);
+	}
+
+	public IndexNode(int pageIndex, BufferManager bm, Split split) {
+		super(pageIndex, bm);
+		setIndexType();
+		insert(0, split.key, split.left);
+		setLink(split.right);
+	}
+
+	@Override
+	public Split insert(int key, int value) {
+
+        /* Early split if node is full.
+         * This is not the canonical algorithm for B+ trees,
+	     * but it is simpler and it does break the definition
+	     * which might result in immature split, which might not be desired in database
+	     * because additional split lead to tree's height increase by 1, thus the number of disk read
+	     * so first search to the leaf, and split from bottom up is the correct approach.
+	     */
+		int size = getSize();
+		if (size == N) { // Split
+            int mid = (N + 1) / 2;
+            int sNum = size - mid;
+            IndexNode sibling = new IndexNode(bm.getFreePage(), bm);
+            
+            for(int j = 0; j < sNum; j ++) {
+            	sibling.insert(j, getKey(mid + j), getData(mid + j));
+            }
+
+            sibling.setLink(getLink());
+            setLink(getData(mid));
+            int removedKey = getKey(mid - 1);
+            setSize(mid - 1);//this is important, so the middle one elevate to next depth(height), inner node's key don't repeat itself
+
+            // Set up the return variable
+            Split result = new Split(removedKey,
+                    this.pageIndex,
+                    sibling.pageIndex);
+
+            // Now insert in the appropriate sibling
+            if (key <= result.key) {
+            	insert(key, value);
+            } else {
+                sibling.insert(key, value);
+            }
+            return result;
+
+        }
+		int index = findKey(key, true);
+		if(index == size) {
+			index = size - 1;
+		}
+
+		int currentKey = getKey(index);
+		int nextNodePage;
+		if(key <= currentKey) {
+			nextNodePage = getData(index);
+		} else {
+			if(index == size - 1) {
+				nextNodePage = getLink();
+			} else {
+				nextNodePage = getData(index + 1);
+			}
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 66:74, 96:104.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.